dask/array/routines.py

```diff
@@ -882,6 +882,120 @@ def histogram(a, bins=None, range=None, normed=False, weights=None, density=None
+        Number of bins in each dimension to histogram.
+    range : sequence of pairs
+        Axes limits in each dimension (xmin, xmax)
+    weights : dask Array, optional
+        Optional weights associated with the data. (TODO)
+    density : bool
+        Convert bins to values of a PDF (TODO)
+
+    See Also
+    --------
+    histogram
+
+    """
+
+    # N == total number of samples
+    # D == total number of dimensions
+    N, D = sample.shape
+
+    # TODO: handle weighted data.
+
+    # generate token and name for task
+    token = tokenize(sample, bins, range, weights, density)
+    name = "histogramdd-sum-" + token
+
+    ######### set up bins #################
+    # In this prototype the bins are _not_ delayed, the use of
+    # _linspace_from_delayed is taking into consideration future
+    # compatibility with delayed bin edge arrays. We're just
+    # converting the pairs of bins and ranges definitions into a
+    # bunch of (immediately available) arrays of bin edges.
+    bins_edges = [
+        _linspace_from_delayed(r[0], r[1], b + 1) for b, r in zip(bins, range)
+    ]
+    ## total number of bin edge arrays must be the total number of dimensions
+    assert len(bins_edges) == D
+    (bins_ref, ranges_ref), deps = unpack_collections([bins_edges, range])
+
+    ######### set up graph ################
+    # New graph where the callable is _block_histogramdd and the
+    # function arguments are chunks of our sample array followed by
+    # the bin information (range information is unncessary since we
+    # converted to arrays of bin edges). The graph is stacked for each
+    # chunk, each building block is a D-dimensional histogram result.
+    dsk = {
+        (name, i, 0): (_block_histogramdd, k, bins_edges)
+        for i, k in enumerate(flatten(sample.__dask_keys__()))
+    }
+    # da.histogram does this to get the dtype (TODO: figure out why)
+    dtype = np.histogramdd([])[0].dtype
+    # da.histogram does this to track possible dependencies from
+    # potentially delayed binning information.
+    final_deps = (sample,) + deps
+    graph = HighLevelGraph.from_collections(name, dsk, dependencies=final_deps)
+
+    ######### set up array ################
+    # this logic yields chunks where the shape is:
+    # (nchunks, nbin_D1, nbin_D2, nbin_D3, ...)
+    # such that we add (collapse) the chunks along axis=0 to get the
+    # final result.
+    #
+    # steps:
+    # 1. get the total number of chunks in the sample
+    # 2. make a tuple of tuples; each inner tuple is the total number
+    #    of bins in a dimension
+    # 3. make "stacked chunks" tuple which combines information from
+    #    steps 1 and 2
+    # 4. make dask Array and collapse the 0th axis (which has size
+    #    nchunks_in_sample)
+    nchunks_in_sample = len(list(flatten(sample.__dask_keys__())))
+    all_nbins = tuple((b.size - 1,) for b in bins_edges)
+    stacked_chunks = ((1,) * nchunks_in_sample, *all_nbins)
```

Comment:
If I change line 989 here to
```python
stacked_chunks = ((1,) * nchunks_in_sample, (D,))
```
(`D` being the dimensions of the histogram) I'm seeing some strange behavior. At first I thought I was on the right track after making this change, because if I call `.compute()` on the output dask Array I actually get the expected resulting numpy array.  But `assert_eq` in the test was failing due to a shape issue. The shape reported by the lazy array is also inconsistent with the shape of the numpy array after calling `compute()` (I'm not sure how that is possible! the graph is also connected in what appears to be a correct way). So even though the actual final piece of numpy data looks correct, it seems that the path to get there is either considered fragile or incorrect.

The line in the PR as it stands:
```python
stacked_chunks = ((1,) * nchunks_in_sample, *all_nbins)
```
which, if `bins = (4, 5, 6)` and `sample` has 3 chunks, would expand to:
```python
((1, 1, 1), (4,), (5,), (6,))
```
is what I _think_ is the correct because the resulting lazy dask Array does report the correct shape (`(4, 5, 6)` for the example here), but fails when calling `.compute`. And the graph from `visualize()` is disconnected (as shown in [my comment here](https://github.com/dask/dask/pull/7346#issuecomment-794186649)).

After this I'm a bit stumped. I'm still tinkering, but I could use a nudge in the right direction if anyone following along might have any tips or ideas

Reply:
Alright the issue had nothing to do with defining the chunks for the output, it was in the definition of the low level graph.